src/version.c

```diff
@@ -74,60 +74,40 @@ version_cmp(VersionInfo *v1, VersionInfo *v2)
 	return 0;
 }
 
-#define NUM_VERSION_DELIMS 4
-
-static const char *const version_delimiter[NUM_VERSION_DELIMS] = {".", ".", "-", ""};
-
 bool
 version_parse(const char *version, VersionInfo *result)
 {
-	char	   *parse_version = pstrdup(version);
-	int			i;
+	size_t		version_len = strlen(version);
+	int			parsed_len = 0;
+	int			fields_parsed = 0;
+	int			i = 0;
 
 	memset(result, 0, sizeof(VersionInfo));
 
-	for (i = 0; i < NUM_VERSION_DELIMS; i++)
-	{
-		const char *subversion;
-
-		subversion = strtok(i == 0 ? parse_version : NULL, version_delimiter[i]);
-
-		if (subversion == NULL)
-			return i > 0;
-
-		/*
-		 * If we are past the '-' delimiter, we've found the mod/pre-release
-		 * tag
-		 */
-		if (i > 0 && version_delimiter[i - 1][0] == '-')
-		{
-			int			len = snprintf(result->version_mod, sizeof(result->version_mod) - 1, "%s", subversion);
-
-			if (len > (sizeof(result->version_mod) - 1))
-				return false;
-
-			if (len > 0)
-				result->has_version_mod = true;
-		}
-		else
-		{
-			char	   *endptr;
-
-			Assert(i < VERSION_PARTS);
-
-			result->version[i] = strtol(subversion, &endptr, 10);
-
-			/*
-			 * We expect the parsing of the version num to end at a '\0' since
-			 * strtok() should have replaced the delimiter with a '\0' if the
-			 * delimiter was found
-			 */
-			if (endptr != NULL && *endptr != '\0')
-				return false;
-		}
-	}
+	/*
+	 * a verion is a string of the form
+	 *
+	 * <number>[.<number>[.<number>[-<string>]]]
+	 *
+	 * after parsing each field in the version-string, we output the number of
+	 * bytes currently parsed, a version-string is valid if all of the bytes
+	 * of the string were parsable using the above grammar.
+	 */
+	fields_parsed = sscanf(version, "%lu%n.%lu%n.%lu%n-%" STR(VERSION_INFO_LEN) "s%n",
+						   &result->version[0], &parsed_len,
+						   &result->version[1], &parsed_len,
+						   &result->version[2], &parsed_len,
+						   result->version_mod, &parsed_len);
```

Comment:
Please double check if passing `parsed_len` multiple times to sscanf is safe.

Reply:
the standard is ambiguous switched to multiple lengths just to be sure.